static/js/choropleth/choropleth.ts

```diff
@@ -0,0 +1,178 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import React from "react";
+import ReactDOM from "react-dom";
+import axios from "axios";
+import { MainPane } from "./choropleth_template";
+
+const d3 = require("d3");
+var url = new URL(window.location.href);
+
+/**
+ * Generates choropleth map from API on pageload.
+ */
+window.onload = () => {
+    const urlParams = new URLSearchParams(window.location.search);
+    ReactDOM.render(
+        React.createElement(MainPane),
+        document.getElementById("main-pane")
+    );
+
+    // Build api call from request.
+    var base_url = "choropleth/download?";
+    base_url += "statVar=" + urlParams.get("statVar");
+    base_url += "&perCapita=" + (urlParams.get("perCapita") == "true");
+    base_url += "&place=" + urlParams.get("place");
+
+    // Create request and generate map.
+    axios.get(base_url).then((resp) => {
+        var payload = resp.data[0];
+        generateGeoMap(payload['geoJson'], payload['_PLOTTING_INFO'])
+    });
+
+    // Generate breadcrumbs.
+    generateBreadCrumbs()
+};
+
+/**
+ * Handles the creation of d3 map from geojson file.
+ * @param {json} geojson entire geoJson map.
+ */
+function generateGeoMap(geojson, plt_info) {
+    // Combine path elements from D3 content.
+    var mapContent = d3.select("#main-pane g.map")
+        .selectAll("path")
+        .data(geojson.features);
+
+    // Build chart display options.
+    var colorScale = d3.scaleLinear()
+      .domain(plt_info['domain'])
+      .range(plt_info['palette']);    
+
+    // Scale and center the map.
+    var svg_container = document.getElementById("map_container")
+    var projection = d3.geoAlbers().fitSize([
+        svg_container.clientWidth,
+        svg_container.clientHeight],
+        geojson);
+    var geomap = d3.geoPath().projection(projection);
+
+    // Build map objects.
+    mapContent.enter()
+        .append("path")
+        .attr("d", geomap)
+        // Add CSS class to each path for border outlining.
+        .attr("class", "border")
+        .attr("fill", function (d) {
+            if (d.properties.hasOwnProperty('value')) {
+                return colorScale(d.properties.value)
+            } else {
+                return "gray";
+            }
+          })
+        // Add various country level event handlers.
+        .on("mouseover", handleMapHover)
+        .on("mouseleave", mouseLeave)
+        .on("click", handleMapClick);
+}
+
+/**
+ * Capture hover event on country and displays relevant information.
+ * @param {json} country is the geoJson content for the hovered country.
+ */
+function handleMapHover(country) {
+    // Display statistical variable information on hover.
+    let name = country.properties.name;
+    if (country.properties.hasOwnProperty('value')) {
+        document.getElementById("display").innerHTML = name + " - " +
```

Comment:
This code block can be much more simplified. The only difference is appending "No value" instead of the value. Can you try to refactor this so there is only one call to set innerHTML?

Reply:
Also, display should get a more descriptive element id. Everything in a UI is a displayed element...

Reply:
Changed both.